platform-core-module/src/main/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleter.java

```diff
@@ -0,0 +1,124 @@
+ *
+ * @author Thees Koester
+ */
+public class ProgramExecutionDeleter {
+
+	private final int throttlingMilliseconds;
+	private final Day referenceDay;
+
+	public ProgramExecutionDeleter(int throttlingMilliseconds) {
+
+		this.throttlingMilliseconds = throttlingMilliseconds;
+		this.referenceDay = Day.today();
+	}
+
+	public void delete() {
+
+		Log.finfo("%s started for %s", ProgramExecutionDeleter.class.getSimpleName(), referenceDay);
+
+		for (AGProgram program: AGProgram.TABLE.createSelect().orderBy(AGProgram.ID)) {
+			deleteExecutionsOfProgram(program);
+		}
+	}
+
+	private void deleteExecutionsOfProgram(AGProgram program) {
+
+		Log.finfo("Cleanup executions of program %s", fetchProgramNameIfPossible(program));
+
+		AGProgramExecution
+			.createSelect()
+			.where(AGProgramExecution.PROGRAM_UUID.equal(program.getProgramUuid()))
+			.where(AGProgramExecution.TERMINATED_AT.isNotNull())
+			.orderBy(AGProgramExecution.TERMINATED_AT)
+			.getFirstAsOptional()
+			.map(AGProgramExecution::getTerminatedAt)
+			.map(DayTime::getDay)
+			.ifPresentOrElse(minimalDayInDb -> deleteBatchwise(program, minimalDayInDb), this::logInfoNothingToDelete);
+	}
+
+	private String fetchProgramNameIfPossible(AGProgram program) {
+
+		try {
+			// Will fail in JUnit tests
+			return program.toDisplayWithoutId() + "";
+		} catch (Exception exception) {
+			DevNull.swallow(exception);
+		}
+		return program + "";
+	}
+
+	private void deleteBatchwise(AGProgram program, Day minimalDayInDb) {
+
+		Day minimalRetainedDay = referenceDay.getRelative(-program.getRetentionDaysOfExecutions());
+
+		if (minimalDayInDb.isBefore(minimalRetainedDay)) {
+			int distance = minimalDayInDb.getDistance(minimalRetainedDay);
+			Day firstThresholdDay = minimalDayInDb.getNext();
+			Log.finfo("Deleting records in %s day-based batches, from %s to %s...", distance, firstThresholdDay, minimalRetainedDay);
+
+			for (Day thresholdDay: new DateItemRange<>(firstThresholdDay, minimalRetainedDay)) {
+				deleteExecutionsOlderThan(program, thresholdDay);
+				sleepIfNecessary(thresholdDay, minimalRetainedDay);
+			}
+		} else {
+			logInfoNothingToDelete();
+		}
+	}
+
+	private void deleteExecutionsOlderThan(AGProgram program, Day thresholdDay) {
+
+		List<AGProgramExecution> programExecutions = loadToBeDeletedProgramExecutions(program, thresholdDay);
+		Log.finfo("Deleting %d program execution(s) older than %s...", programExecutions.size(), thresholdDay);
+		try (DbTransaction transaction = new DbTransaction()) {
+			deleteProgramExecutionLogsAndProgramExecutions(programExecutions);
+			transaction.commit();
+		}
+	}
+
+	private List<AGProgramExecution> loadToBeDeletedProgramExecutions(AGProgram program, Day thresholdDay) {
+
+		return AGProgramExecution.TABLE
+			.createSelect()
+			.where(AGProgramExecution.PROGRAM_UUID.equal(program.getProgramUuid()))
+			.where(AGProgramExecution.TERMINATED_AT.isNotNull())
```

Comment:
I think this line can be omitted because, in SQL, a comparison with `NULL`
- does not cause an exception, and
- always yields `false` (so that records with `TERMINATED_AT = NULL` will never be in the result set).

Reply:
done